components/server/src/workspace/image-source-provider.ts

```diff
@@ -37,17 +45,26 @@ export class ImageSourceProvider {
                 }
             } else if (ImageConfigFile.is(imgcfg)) {
                 // There are no special instructions as to where to get the Dockerfile from, hence we use the context of the current workspace.
+                if (AdditionalContentContext.is(context) && ImageConfigFile.is(config.image) && context.additionalFiles[config.image.file]) {
+                    return {
+                        dockerFilePath: config.image.file,
+                        dockerFileContentSHA: this.getContentSHA(context.additionalFiles[config.image.file]),
+                    }
+                }
                 const hostContext = this.hostContextProvider.get(context.repository.host);
                 if (!hostContext || !hostContext.services) {
                     throw new Error(`Cannot fetch workspace image source for host: ${context.repository.host}`);
                 }
-                const lastDockerFileSha = await hostContext.services.fileProvider.getLastChangeRevision(context.repository, context.revision, user, imgcfg.file);
+                const lastDockerFileShaPromise = hostContext.services.fileProvider.getLastChangeRevision(context.repository, context.revision, user, imgcfg.file);
```

Comment:
we need to make two requests for the time being, to support the current resolution of image refs. Once we have rebuilt most images, we can remove the `lastChangeRevision` part.

Reply:
If we resolved the base image here already we could check which case (`lastDockerFileSha` or `this.getContentSHA(dockerFileContent)`) resolves to a built image. This behaviour is what I referred to as "move the legacy behaviour to server" above.